File: tst/shot.c
#include <SDL2/SDL.h>

#include "pico.h"
#include "../check.h"

void check (const char* out, const char* asr) {
    printf("Testing: %s\n", asr);
    assert(_pico_cmp_files(out, asr));
}

int main (void) {
    pico_init(1);
    pico_output_clear();

    // PICO_OUTPUT_SCREENSHOT

    {
        puts("entire screen - 01");
        Pico_Rel_Rect r = { '!', {10, 10, 20, 20}, PICO_ANCHOR_NW };
        pico_output_draw_rect(r);
        pico_set_layer("window");
        const char* f = pico_output_screenshot(NULL, NULL, NULL);
        pico_set_layer("world");
        assert(f != NULL);
        check(f, "asr/shot-01.png");
        assert(remove(f) == 0);
    }

    {
        puts("entire screen - 02");
        pico_set_pencil_color((Pico_Color){200, 0, 0, 0xFF});
        Pico_Rel_Rect r = { '!', {30, 30, 10, 10}, PICO_ANCHOR_NW };
        pico_output_draw_rect(r);
        pico_set_layer("window");
        const char* f = pico_output_screenshot(NULL, "out/shot-02.png", NULL);
        pico_set_layer("world");
        assert(!strcmp(f, "out/shot-02.png"));
        check(f, "asr/shot-02.png");
    }

    {
        puts("part of screen (raw)");
        pico_set_pencil_color((Pico_Color){0, 200, 0, 0xFF});
        Pico_Rel_Rect r = { '!', {40, 5, 10, 10}, PICO_ANCHOR_NW };
        pico_output_draw_rect(r);
        Pico_Rel_Rect clip = { '!', {0, 0, 250, 150}, PICO_ANCHOR_NW };
        pico_set_layer("window");
        const char* f = pico_output_screenshot(NULL, NULL, &clip);
        pico_set_layer("world");
        assert(f != NULL);
        check(f, "asr/shot-03.png");
        assert(remove(f) == 0);
    }

    {
        puts("part of screen (pct)");
        pico_set_pencil_color((Pico_Color){0, 0, 200, 0xFF});
        pico_output_draw_rect (
            (Pico_Rel_Rect) { '!', {50, 50, 10, 10}, PICO_ANCHOR_NW }
        );
        pico_set_layer("window");
        const char* f = pico_output_screenshot(NULL, NULL,
            &(Pico_Rel_Rect) { '%', {0, 0, 0.5, 0.3}, PICO_ANCHOR_NW }
        );
        pico_set_layer("world");
        assert(f != NULL);
        check(f, "asr/shot-04.png");
        assert(remove(f) == 0);
    }

    {
        puts("world layer");
        pico_set_effect_color((Pico_Color){0x10, 0x10, 0x10, 0xFF});
        pico_output_clear();
        pico_set_pencil_color((Pico_Color){0xFF, 0x00, 0x00, 0xFF});
        pico_output_draw_rect (
            (Pico_Rel_Rect) { '!', {10, 10, 30, 30}, PICO_ANCHOR_NW }
        );
        const char* f = pico_output_screenshot(NULL, "out/shot-world.png", NULL);
        assert(!strcmp(f, "out/shot-world.png"));
        check(f, "asr/shot-world.png");
    }

    {
        puts("empty layer");
        pico_layer_empty(NULL, "empty1", 1, (Pico_Rel_Dim){'!', {64, 32}}, NULL);
        pico_set_layer("empty1");
        pico_set_effect_color((Pico_Color){0x00, 0x80, 0x00, 0xFF});
        pico_output_clear();
        pico_set_pencil_color((Pico_Color){0xFF, 0xFF, 0x00, 0xFF});
        pico_output_draw_rect (
            (Pico_Rel_Rect) { '!', {5, 5, 10, 10}, PICO_ANCHOR_NW }
        );
        const char* f = pico_output_screenshot(NULL, "out/shot-empty.png", NULL);
        pico_set_layer("world");
        assert(!strcmp(f, "out/shot-empty.png"));
        check(f, "asr/shot-empty.png");
    }

    {
        puts("pixmap layer");
        static Pico_Color buf[4] = {
            {255,   0,   0, 255}, {  0, 255,   0, 255},
            {  0,   0, 255, 255}, {255, 255,   0, 255}
        };
        pico_layer_pixmap(NULL, "pmap1", (Pico_Abs_Dim){2, 2}, buf);
        pico_set_layer("pmap1");
        const char* f = pico_output_screenshot(NULL, "out/shot-pixmap.png", NULL);
        pico_set_layer("world");
        assert(!strcmp(f, "out/shot-pixmap.png"));
        check(f, "asr/shot-pixmap.png");
    }

    {
        puts("sub layer");
        pico_layer_sub(NULL, "sub1", "empty1",
            &(Pico_Rel_Rect) { '!', {0, 0, 32, 16}, PICO_ANCHOR_NW }
        );
        pico_set_layer("sub1");
        const char* f = pico_output_screenshot(NULL, "out/shot-sub.png", NULL);
        pico_set_layer("world");
        assert(!strcmp(f, "out/shot-sub.png"));
        check(f, "asr/shot-sub.png");
    }

    {
        puts("image layer");
        pico_layer_image(NULL, "img1", "../res/open.png");
        pico_set_layer("img1");
        const char* f = pico_output_screenshot(NULL, "out/shot-image.png", NULL);
        pico_set_layer("world");
        assert(!strcmp(f, "out/shot-image.png"));
        check(f, "asr/shot-image.png");
    }

    {
        puts("text layer");
        pico_set_pencil_color((Pico_Color){0xFF, 0xFF, 0xFF, 0xFF});
        pico_layer_text(NULL, "txt1", 16, "hello");
        pico_set_layer("txt1");
        const char* f = pico_output_screenshot(NULL, "out/shot-text.png", NULL);
        pico_set_layer("world");
        assert(!strcmp(f, "out/shot-text.png"));
        check(f, "asr/shot-text.png");
    }

    {
        puts("video layer");
        pico_layer_video(NULL, "vid1", "video.y4m");
        pico_set_layer("vid1");
        assert(pico_set_video("vid1", 0) == 1);
        const char* f = pico_output_screenshot(NULL, "out/shot-video.png", NULL);
        pico_set_layer("world");
        assert(!strcmp(f, "out/shot-video.png"));
        check(f, "asr/shot-video.png");
    }

    // PICO_LAYER_SCREENSHOT

    // 1. capture an existing layer into a new layer; the new layer must
    // hold the same pixels, so its screenshot matches empty1's reference
    {
        puts("screenshot layer - reuse empty1");
        pico_layer_screenshot(NULL, "snap_empty", "empty1", NULL);
        pico_set_layer("snap_empty");
        const char* f = pico_output_screenshot(NULL, "out/shot-snap-empty.png", NULL);
        pico_set_layer("world");
        assert(!strcmp(f, "out/shot-snap-empty.png"));
        check(f, "asr/shot-empty.png");
    }

    // 2. compose a multi-layer scene on world, capture it as one layer,
    // clear, redraw the capture, and confirm it matches the original scene
    {
        puts("screenshot layer - scene round-trip");

        pico_layer_empty(NULL, "red", 1, (Pico_Rel_Dim){'!', {20, 20}}, NULL);
        pico_set_layer("red");
        pico_set_effect_color((Pico_Color){0xFF, 0, 0, 0xFF});
        pico_output_clear();

        pico_layer_empty(NULL, "blue", 1, (Pico_Rel_Dim){'!', {20, 20}}, NULL);
        pico_set_layer("blue");
        pico_set_effect_color((Pico_Color){0, 0, 0xFF, 0xFF});
        pico_output_clear();

        pico_set_layer("world");
        pico_set_effect_color((Pico_Color){0x20, 0x20, 0x20, 0xFF});
        pico_output_clear();
        pico_output_draw_layer("red", &(Pico_Rel_Rect){ '!', {20, 20, 20, 20}, PICO_ANCHOR_NW });
        pico_output_draw_layer("blue", &(Pico_Rel_Rect){ '!', {50, 40, 20, 20}, PICO_ANCHOR_NW });

        const char* f1 = pico_output_screenshot("world", "out/shot-scene.png", NULL);
        check(f1, "asr/shot-scene.png");

        pico_layer_screenshot(NULL, "snap_scene", "world", NULL);
        pico_set_layer("world");
        pico_output_clear();
        pico_output_draw_layer("snap_scene", NULL);
        const char* f2 = pico_output_screenshot("world", "out/shot-scene-2.png", NULL);
        check(f2, "asr/shot-scene.png");
        pico_set_layer("world");
    }

    // 3. capture a relative region of a layer; the rect resolves against
    // the source layer (empty1 = 64x32), so this yields its left half
    {
        puts("screenshot layer - region (pct)");
        pico_layer_screenshot(NULL, "snap_half", "empty1",
            &(Pico_Rel_Rect){ '%', {0, 0, 0.5, 1.0}, PICO_ANCHOR_NW }
        );
        pico_set_layer("snap_half");
        const char* f = pico_output_screenshot(NULL, "out/shot-snap-half.png", NULL);
        pico_set_layer("world");
        assert(!strcmp(f, "out/shot-snap-half.png"));
        check(f, "asr/shot-snap-half.png");
    }

    // 4. regression: a capture must not be corrupted by a present fired from
    // the event loop during the capture's internal delay. A queued resize
    // event makes _pico_shot's pico_input_delay present, which rebinds the
    // render target to the *current* layer; the capture must still read the
    // requested layer (else it returns a blank/wrong screenshot).
    {
        puts("screenshot layer - render-target race");
        pico_layer_empty(NULL, "race_cur", 1,
            (Pico_Rel_Dim){'!', {64, 32}}, NULL);
        pico_set_layer("race_cur");
        pico_set_effect_color((Pico_Color){0x00, 0x80, 0x00, 0xFF});
        pico_output_clear();        // current layer: solid green, != empty1

        // queue an event the capture's internal pico_input_delay will pick up;
        // resizing to the current size keeps the window content unchanged
        SDL_Event e = {0};
        e.type = SDL_WINDOWEVENT;
        e.window.event = SDL_WINDOWEVENT_RESIZED;
        e.window.data1 = PICO_DIM_PHY.w;
        e.window.data2 = PICO_DIM_PHY.h;
        SDL_PushEvent(&e);

        // capture empty1 while race_cur is current: must match empty1's
        // reference, not race_cur's blank/green content
        const char* f = pico_output_screenshot("empty1", "out/shot-race.png", NULL);
        pico_set_layer("world");
        assert(!strcmp(f, "out/shot-race.png"));
        check(f, "asr/shot-empty.png");
    }

    pico_init(0);
    return 0;
}

File: lua/tst/shot.lua
require 'pico.check'

local function check (out, asr)
    print("Testing: " .. asr)
    local f1 = io.open(out, "rb")
    local f2 = io.open(asr, "rb")
    assert(f1, "cannot open " .. out)
    assert(f2, "cannot open " .. asr)
    local c1 = f1:read("*a")
    local c2 = f2:read("*a")
    f1:close()
    f2:close()
    assert(c1 == c2, "files differ: " .. out .. " vs " .. asr)
end

pico.init(true)
pico.output.clear()

do
    print("entire screen - 01")
    pico.output.draw.rect({'!', x=10, y=10, w=20, h=20, anchor='NW'})
    pico.set.layer("window")
    local f = pico.output.screenshot(nil)
    pico.set.layer("world")
    assert(f ~= nil)
    check(f, "../../tst/asr/shot-01.png")
    assert(os.remove(f))
end

do
    print("entire screen - 02")
    pico.set.pencil { color={'!', r=200, g=0, b=0} }
    pico.output.draw.rect({'!', x=30, y=30, w=10, h=10, anchor='NW'})
    pico.set.layer("window")
    local f = pico.output.screenshot(nil, "../../tst/out/shot-02.png")
    pico.set.layer("world")
    assert(f == "../../tst/out/shot-02.png")
    check(f, "../../tst/asr/shot-02.png")
end

do
    print("part of screen (raw)")
    pico.set.pencil { color={'!', r=0, g=200, b=0} }
    pico.output.draw.rect({'!', x=40, y=5, w=10, h=10, anchor='NW'})
    pico.set.layer("window")
    local f = pico.output.screenshot(nil, {'!', x=0, y=0, w=250, h=150, anchor='NW'})
    pico.set.layer("world")
    assert(f ~= nil)
    check(f, "../../tst/asr/shot-03.png")
    assert(os.remove(f))
end

do
    print("part of screen (pct)")
    pico.set.pencil { color={'!', r=0, g=0, b=200} }
    pico.output.draw.rect({'!', x=50, y=50, w=10, h=10, anchor='NW'})
    pico.set.layer("window")
    local f = pico.output.screenshot(nil, {'%', x=0, y=0, w=0.5, h=0.3, anchor='NW'})
    pico.set.layer("world")
    assert(f ~= nil)
    check(f, "../../tst/asr/shot-04.png")
    assert(os.remove(f))
end

do
    print("world layer")
    pico.set.effect { color={'!', r=0x10, g=0x10, b=0x10} }
    pico.output.clear()
    pico.set.pencil { color={'!', r=0xFF, g=0x00, b=0x00} }
    pico.output.draw.rect({'!', x=10, y=10, w=30, h=30, anchor='NW'})
    local f = pico.output.screenshot(nil, "../../tst/out/shot-world.png")
    assert(f == "../../tst/out/shot-world.png")
    check(f, "../../tst/asr/shot-world.png")
end

do
    print("empty layer")
    pico.layer.empty { key="empty1", clear=false, dim={'!', w=64, h=32} }
    pico.set.layer("empty1")
    pico.set.effect { color={'!', r=0x00, g=0x80, b=0x00} }
    pico.output.clear()
    pico.set.pencil { color={'!', r=0xFF, g=0xFF, b=0x00} }
    pico.output.draw.rect({'!', x=5, y=5, w=10, h=10, anchor='NW'})
    local f = pico.output.screenshot(nil, "../../tst/out/shot-empty.png")
    pico.set.layer("world")
    assert(f == "../../tst/out/shot-empty.png")
    check(f, "../../tst/asr/shot-empty.png")
end

do
    print("pixmap layer")
    pico.layer.pixmap { key="pmap1", pixels={
        {{r=255, g=  0, b=  0, a=255}, {r=  0, g=255, b=  0, a=255}},
        {{r=  0, g=  0, b=255, a=255}, {r=255, g=255, b=  0, a=255}},
    } }
    pico.set.layer("pmap1")
    local f = pico.output.screenshot(nil, "../../tst/out/shot-pixmap.png")
    pico.set.layer("world")
    assert(f == "../../tst/out/shot-pixmap.png")
    check(f, "../../tst/asr/shot-pixmap.png")
end

do
    print("sub layer")
    pico.layer.sub { key="sub1", sup="empty1",
        crop={'!', x=0, y=0, w=32, h=16, anchor='NW'} }
    pico.set.layer("sub1")
    local f = pico.output.screenshot(nil, "../../tst/out/shot-sub.png")
    pico.set.layer("world")
    assert(f == "../../tst/out/shot-sub.png")
    check(f, "../../tst/asr/shot-sub.png")
end

do
    print("image layer")
    pico.layer.image { key="img1", path="../../res/open.png" }
    pico.set.layer("img1")
    local f = pico.output.screenshot(nil, "../../tst/out/shot-image.png")
    pico.set.layer("world")
    assert(f == "../../tst/out/shot-image.png")
    check(f, "../../tst/asr/shot-image.png")
end

do
    print("text layer")
    pico.set.pencil { color={'!', r=0xFF, g=0xFF, b=0xFF} }
    pico.layer.text { key="txt1", height=16, text="hello" }
    pico.set.layer("txt1")
    local f = pico.output.screenshot(nil, "../../tst/out/shot-text.png")
    pico.set.layer("world")
    assert(f == "../../tst/out/shot-text.png")
    check(f, "../../tst/asr/shot-text.png")
end

do
    print("video layer")
    pico.layer.video { key="vid1", path="video.y4m" }
    pico.set.layer("vid1")
    assert(pico.set.video("vid1", 0) == true)
    local f = pico.output.screenshot(nil, "../../tst/out/shot-video.png")
    pico.set.layer("world")
    assert(f == "../../tst/out/shot-video.png")
    check(f, "../../tst/asr/shot-video.png")
end

-- PICO.LAYER.SCREENSHOT

-- 1. capture an existing layer into a new layer; reuses empty1's reference
do
    print("screenshot layer - reuse empty1")
    pico.layer.screenshot { key="snap_empty", sup="empty1" }
    pico.set.layer("snap_empty")
    local f = pico.output.screenshot(nil, "../../tst/out/shot-snap-empty.png")
    pico.set.layer("world")
    assert(f == "../../tst/out/shot-snap-empty.png")
    check(f, "../../tst/asr/shot-empty.png")
end

-- 2. compose a multi-layer scene, capture it, clear, redraw, compare
do
    print("screenshot layer - scene round-trip")

    pico.layer.empty { key="red", dim={'!', w=20, h=20} }
    pico.set.layer("red")
    pico.set.effect { color={'!', r=0xFF, g=0, b=0} }
    pico.output.clear()

    pico.layer.empty { key="blue", dim={'!', w=20, h=20} }
    pico.set.layer("blue")
    pico.set.effect { color={'!', r=0, g=0, b=0xFF} }
    pico.output.clear()

    pico.set.layer("world")
    pico.set.effect { color={'!', r=0x20, g=0x20, b=0x20} }
    pico.output.clear()
    pico.output.draw.layer("red", {'!', x=20, y=20, w=20, h=20, anchor='NW'})
    pico.output.draw.layer("blue", {'!', x=50, y=40, w=20, h=20, anchor='NW'})

    local f1 = pico.output.screenshot("world", "../../tst/out/shot-scene.png")
    check(f1, "../../tst/asr/shot-scene.png")

    pico.layer.screenshot { key="snap_scene", sup="world" }
    pico.set.layer("world")
    pico.output.clear()
    pico.output.draw.layer("snap_scene")
    local f2 = pico.output.screenshot("world", "../../tst/out/shot-scene-2.png")
    check(f2, "../../tst/asr/shot-scene.png")
    pico.set.layer("world")
end

-- 3. capture a relative crop; resolves against the source layer (64x32)
do
    print("screenshot layer - crop (pct)")
    pico.layer.screenshot { key="snap_half", sup="empty1",
        crop={'%', x=0, y=0, w=0.5, h=1.0, anchor='NW'} }
    pico.set.layer("snap_half")
    local f = pico.output.screenshot(nil, "../../tst/out/shot-snap-half.png")
    pico.set.layer("world")
    assert(f == "../../tst/out/shot-snap-half.png")
    check(f, "../../tst/asr/shot-snap-half.png")
end

-- 4. constructor target: screenshot layer placed via target field,
--    combined with crop; stored as scene.target (raw), then drawn
--    with no rect (uses the target) into world's SE quadrant
do
    print("screenshot layer - target (+ crop)")
    local r = {'%', x=1, y=1, w=0.5, h=0.5, anchor='SE'}
    pico.layer.screenshot { key="snap_tgt", sup="empty1",
        crop={'%', x=0, y=0, w=0.5, h=1.0, anchor='NW'}, target=r }
    pico.set.layer("snap_tgt")
    local s = pico.get.scene()
    assert(s.target[1] == '%')
    assert(s.target.x == 1 and s.target.y == 1)
    assert(s.target.w == 0.5 and s.target.h == 0.5)
    pico.set.layer("world")
    pico.set.effect { color='black' }
    pico.output.clear()
    pico.output.draw.layer("snap_tgt")
    local f = pico.output.screenshot("world", "../../tst/out/shot-snap-target.png")
    assert(f == "../../tst/out/shot-snap-target.png")
    check(f, "../../tst/asr/shot-snap-target.png")
end

pico.init(false)
